pkg/containerd/cmd/service/start.go

```diff
@@ -0,0 +1,123 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/containerd/containerd"
+	"github.com/containerd/containerd/api/services/containers"
+	"github.com/containerd/containerd/api/services/execution"
+	//"github.com/containerd/containerd/api/types/mount"
+	"github.com/containerd/containerd/namespaces"
+	specs "github.com/opencontainers/runtime-spec/specs-go"
+)
+
+func start(args []string) {
+	invoked := filepath.Base(os.Args[0])
+	flags := flag.NewFlagSet("start", flag.ExitOnError)
+	flags.Usage = func() {
+		fmt.Printf("USAGE: %s start [service]\n\n", invoked)
+		fmt.Printf("Options:\n")
+		flags.PrintDefaults()
+	}
+
+	sock := flags.String("sock", "/run/containerd/containerd.sock", "Path to containerd socket")
+
+	dumpSpec := flags.String("dump-spec", "", "Dump container spec to file before start")
+
+	if err := flags.Parse(args); err != nil {
+		log.Fatal("Unable to parse args")
+	}
+	args = flags.Args()
+
+	if len(args) != 1 {
+		fmt.Println("Please specify the service")
+		flags.Usage()
+		os.Exit(1)
+	}
+
+	service := args[0]
+	rootfs := filepath.Join("/containers/services", service, "rootfs")
+	log.Infof("Starting service: %q", service)
+	log := log.WithFields(log.Fields{
+		"service": service,
+	})
+
+	client, err := containerd.New(*sock)
+	if err != nil {
+		log.Fatal("creating containerd client")
+	}
+
+	ctx := namespaces.WithNamespace(context.Background(), "default")
+
+	var spec *specs.Spec
+	specf, err := os.Open(filepath.Join("/containers/services", service, "config.json"))
+	if err != nil {
+		log.Fatalf("failed to read config.json: %s", err)
+	}
+	if err := json.NewDecoder(specf).Decode(&spec); err != nil {
+		log.Fatalf("failed to parse config.json: %s", err)
+	}
+
+	log.Debugf("Rootfs is %s", rootfs)
+
+	spec.Root.Path = rootfs
+
+	if *dumpSpec != "" {
+		d, err := os.Create(*dumpSpec)
+		if err != nil {
+			log.Fatalf("opening file for spec dump: %s", err)
+		}
+		enc := json.NewEncoder(d)
+		enc.SetIndent("", "    ")
+		if err := enc.Encode(&spec); err != nil {
+			log.Fatalf("dumping spec: %s", err)
+		}
+
+	}
+
+	ctr, err := client.NewContainer(ctx, service, containerd.WithSpec(spec),
+		func(ctx context.Context, client *containerd.Client, c *containers.Container) error {
+			c.RootFS = service // This seems odd, but mimics ctr.
```

Comment:
A bit more experimentation suggests that in this mode of operation not initialising `c.RootFS` is fine (and correct).